### PR TITLE
RFC: Various Nullable comparison methods

### PR DESCRIPTION
**Pull request by davidanthoff**
This PR makes two changes:

1. It adds mixed ``Nullable`` and non-``Nullable`` ``isequal`` and ``isless`` methods.
2. It adds methods to ``==`` that handle ``Nullable``s. The semantics of ``==`` are identical to the semantics of ``isequal`` for the ``Nullable`` case (which is also the C# semantics).

This is essentially an implementation of my suggestion from #20504.

I still need to write tests and probably more documentation, but before I put more work into this I'd like to get feedback whether the core group would consider merging this.

In terms of timing, I assume this would go in after 0.6 has branched.

Here are some arguments why I think it would be good to merge this:

1. I could move [Query.jl](https://github.com/davidanthoff/Query.jl) back to using ``Nullable``. Now, maybe I'm the only one who cares, but I do think that would be valuable.
2. I think adding support for ``==(x::Nullable, y::Nullable)`` makes things a lot more consistent. All the other comparison operators, like ``<`` etc. already work for ``Nullable``s, and it seems odd that ``==`` is the only comparison operator that throws an error.
3. I don't really see a big drawback in adding the mixed ``Nullable`` and non-``Nullable`` operators, but the payoff in terms of usability seems large to me.

CC @nalimilan

### CODE REVIEW

**Comment by davidanthoff** [comment]
I think I have a handle on the test failures, I'll try to push something tomorrow addressing those.

**Comment by davidanthoff** [comment]
Alright, I had to add ``=={T}(x::Nullable{T}, y::WeakRef)`` and ``=={T}(x::WeakRef, y::Nullable{T})`` to get rid of some ambiguity errors, now everything should work.

File: base/nullable.jl [hunk 1]
```diff
@@ -210,11 +210,66 @@ and `false` if one is null but not the other: nulls are considered equal.
...
 
+@inline function isequal{S,T}(x::Nullable{S}, y::T)
+    if null_safe_op(isequal, S, T)
+        !isnull(x) & isequal(x.value, y)
+    else
+        !isnull(x) && isequal(x.value, y)
+    end
+end
+
+@inline function isequal{S,T}(x::S, y::Nullable{T})
```
Comment (nalimilan) [inline base/nullable.jl]: This can simply call `isequal(y, x)`. Same for `isless`.

**Comment by nalimilan** [comment]
While I agree these semantics make sense, I think we should wait until we have decided what representation of nullable we want to retain for 1.0: better have the whole picture clear before making such changes. In particular, if we decide to have two different kinds of nullable types, one for missing values and one more like an `Optional` type, only the former would support operations mixing nullables and scalars.

**Comment by ararslan** [comment]
To expand on Milan's comment above, since we're in feature freeze for Julia 0.6, this would have to wait until the next dev cycle to be merged. In said dev cycle, Jameson is going to unleash his optimizations for `Union`s, which will make the scalar behavior as used in DataArrays much faster. As I understand it, the hope is then to gut the Base `Nullable` type and replace it with two things: the "data analyst's null," which is just a simple `Union{T, Null}`, and the "software engineer's null," which will be a container. The latter is akin to Rust's `Option` type while the former is akin to R's `NA`.

**Comment by TotalVerb** [comment]
We need to be careful with hashing too, with the `isequal` change.

**Comment by davidanthoff** [comment]
The last time I had checked into the whole ``Nullable`` debate I had gotten the impression that folks were moving away from the idea of two separate types? Can you point me towards the discussion where that took another turn?

Also, is there some discussion of the new plan that @ararslan mentions? For example, will we be able to construct efficient structs that have ``Union`` type fields?

**Comment by TotalVerb** [comment]
Right, I was also unaware of the plan to move toward two separate types. A link to the discussion would be helpful.

**Comment by ararslan** [comment]
There have been many related discussions across various mediums, including the DataFrames, DataTables, NullableArrays, Julia, and Juleps repositories on GitHub, John Myles White's meeting with Eric Lippert and Erik Meijer to discuss the relevant decisions made in the design of C#, and offline discussions amongst various parties. No formal decision has been made by any means; the idea to have two nulls that serve mutually exclusive purposes was more of a consensual feeling that has grown from the many discussions relating to general nullables.

> will we be able to construct efficient structs that have Union type fields?

It is my understanding that with Jameson's planned optimizations for `Union`s in the next development cycle, that will indeed be the case.

**Comment by davidanthoff** [comment]
Alright, for now I'm moving the ``DataValue`` type from [Query.jl](https://github.com/davidanthoff/Query.jl) into its own package [DataValues.jl](https://github.com/davidanthoff/DataValues.jl). ``DataValue`` is essentially identical to ``Nullable``, plus things similar to this PR and a few other things that make working with missing data easier.

At least for [Query.jl](https://github.com/davidanthoff/Query.jl) that seems the best solution for now, i.e. I'll use that type for now, and if something materializes in base for 1.0 I can still re-evaluate.

**Pull request closed without merging**